crates/hotshot-state-prover/src/circuit.rs

```diff
@@ -116,236 +109,244 @@ impl AsRef<[Variable]> for LightClientStateVar {
+                    signer_bit_vec_var[2 * i].0,
+                    stake_table_var[2 * i + 1].stake_amount,
+                    signer_bit_vec_var[2 * i + 1].0,
+                ],
+                &[F::one(), F::one()],
+            )
+        })
+        .collect::<Result<Vec<_>, CircuitError>>()?;
+    // Adding the last if STAKE_TABLE_CAPACITY is not a multiple of 2
+    if STAKE_TABLE_CAPACITY % 2 == 1 {
+        signed_amount_var.push(circuit.mul(
+            stake_table_var[STAKE_TABLE_CAPACITY - 1].stake_amount,
+            signer_bit_vec_var[STAKE_TABLE_CAPACITY - 1].0,
+        )?);
     }
+    let acc_amount_var = circuit.sum(&signed_amount_var)?;
+    circuit.enforce_leq(threshold_pub_var, acc_amount_var)?;
+
+    // checking the commitment for the list of schnorr keys
+    let schnorr_ver_key_preimage_vars = stake_table_var
+        .iter()
+        .flat_map(|var| [var.schnorr_ver_key.0.get_x(), var.schnorr_ver_key.0.get_y()])
+        .collect::<Vec<_>>();
+    let schnorr_ver_key_comm = RescueNativeGadget::<F>::rescue_sponge_with_padding(
+        &mut circuit,
+        &schnorr_ver_key_preimage_vars,
+        1,
+    )?[0];
+    circuit.enforce_equal(
+        schnorr_ver_key_comm,
+        lightclient_state_pub_var
+            .stake_table_comm()
+            .schnorr_keys_comm,
+    )?;
+
+    // checking the commitment for the list of stake amounts
+    let stake_amount_preimage_vars = stake_table_var
+        .iter()
+        .map(|var| var.stake_amount)
+        .collect::<Vec<_>>();
+    let stake_amount_comm = RescueNativeGadget::<F>::rescue_sponge_with_padding(
+        &mut circuit,
+        &stake_amount_preimage_vars,
+        1,
+    )?[0];
+    circuit.enforce_equal(
+        stake_amount_comm,
+        lightclient_state_pub_var
+            .stake_table_comm()
+            .stake_amount_comm,
+    )?;
+
+    // checking all signatures
+    let verification_result_vars = stake_table_var
+        .iter()
+        .zip(sig_vars)
+        .map(|(entry, sig)| {
+            SignatureGadget::<_, P>::check_signature_validity(
+                &mut circuit,
+                &entry.schnorr_ver_key,
+                lightclient_state_pub_var.as_ref(),
+                &sig,
+            )
+        })
+        .collect::<Result<Vec<_>, CircuitError>>()?;
+    let bit_x_result_vars = signer_bit_vec_var
+        .iter()
+        .zip(verification_result_vars)
+        .map(|(&bit, result)| {
+            let neg_bit = circuit.logic_neg(bit)?;
+            circuit.logic_or(neg_bit, result)
+        })
+        .collect::<Result<Vec<_>, CircuitError>>()?;
+    let sig_ver_result = circuit.logic_and_all(&bit_x_result_vars)?;
+    circuit.enforce_true(sig_ver_result.0)?;
+
+    circuit.finalize_for_arithmetization()?;
+    Ok((circuit, public_inputs))
 }
 
 #[cfg(test)]
 mod tests {
-    use super::{LightClientState, StateUpdateBuilder};
+    use super::{build_state_verifier_circuit, LightClientState};
```

Comment:
Should add `From<LightClientState> for Vec<F>` or maybe `LightClient::to_fields()` so that we don't need to manually do `let state_msg = [ /* manually constructing each items */ ]` later in the test

Reply:
Implements a `to_array()` for `LightClientState` in 88784b9

Reply:
may I ask why not `to_fields`? that seems clearer than just generic "array"

Reply:
That's just my preference of array over vector. Since it's fixed length.